### PR TITLE
Move the recalculateZoom to the end of the movement.

### LINKED ISSUE #4125: Hash is not synced with map position when terrain is enabled.
**maplibre-gl-js version**: 4.3.1

**browser**: Chrome

### Steps to Trigger Behavior

 1. `npm run start`
 2. `http://localhost:9966/test/examples/3d-terrain.html`
 3. Move the map a bit
 4. Wait 1 second
 5. In the console of the browser execute `map.getZoom()`
 6. Compare the zoom value to the addressbar hash value - they are different buy just a bit.

### Link to Demonstration

image

### Expected Behavior

Zoom in the address bar and in the console (current map state) should be the same after the map had stopped moving.

### Actual Behavior

The zoom values are different.

### PR DESCRIPTION
**Pull request by HarelM**
## Launch Checklist

- #4125

Fixes #4125 - by moving the recalculate zoom from the moveend event handler to the end of the movement.

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to related issues.
 - [x] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality.
 - [x] Document any changes to public APIs.
 - [x] Post benchmark scores.
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

### CODE REVIEW

**Comment by HarelM** [comment]
@olsen232 would be great to have your input here.

**Comment by olsen232** [comment]
I'm on holiday until June 8... just had a tiny look at this, didn't immediately understand it. I don't really have any wisdom about how this code should look, I've just been identifying and fixing bugs case by case

File: src/ui/handler_manager.ts [hunk 1]
```diff
@@ -590,7 +585,13 @@ export class HandlerManager {
         }
 
         const stillMoving = isMoving(this._eventsInProgress);
-        if (allowEndAnimation && (wasMoving || nowMoving) && !stillMoving) {
+        const finishedMoving = (wasMoving || nowMoving) && !stillMoving;
+        if (finishedMoving && this._terrainMovement) {
```
Comment (sbachinin) [inline src/ui/handler_manager.ts:589]: allowEndAnimation check is not needed here?

**Review by sbachinin — APPROVED** [review]

File: src/ui/handler_manager.ts (hunk 1 above)
Comment (HarelM) [reply src/ui/handler_manager.ts:589]: No, we need to adjust the zoom regardless of the animation flag.

Comment (HarelM) [reply src/ui/handler_manager.ts:589]: I think...

**Comment by HarelM** [comment]
@olsen232 thanks for looking into it. 
I'll merge this and hopefully most of the zoom related stuff will be solved after this. 
If you happen to find other bugs feel free to tackle them.
Cc: @prozessor13

**Pull request merged**